File: outflo-assignment-backend/scraper.js
const puppeteer = require("puppeteer");
const { MongoClient } = require("mongodb");

const mongoUri =
  "mongodb+srv://mavinash422:cCRAQrT8blgY5fWf@cluster0.bic32gr.mongodb.net/Outflo?retryWrites=true&w=majority&appName=Cluster0";
const client = new MongoClient(mongoUri);
const dbName = "Outflo";

const scrapeLinkedIn = async () => {
  const browser = await puppeteer.launch({
    headless: false,
    userDataDir: "./user_data",
    defaultViewport: null,
    dumpio: true, // 🐞 Log browser output to terminal
    slowMo: 50, // 🐢 Slow down actions (optional for debugging)
    args: [
      "--disable-blink-features=AutomationControlled", // 🕵️‍♂️ Less detectable
    ],
  });

  const page = await browser.newPage();
  const searchUrl =
    "https://www.linkedin.com/search/results/people/?geoUrn=%5B%22103644278%22%5D&industry=%5B%221594%22%2C%221862%22%2C%2280%22%5D&keywords=%22lead%20generation%20agency%22&origin=GLOBAL_SEARCH_HEADER&sid=z%40k&titleFreeText=Founder";

  try {
    console.log("Step 0: Going to search URL...");
    await page.goto(searchUrl, { timeout: 60000 });
    console.log("Step 1: Page loaded.");

    // Wait for 5 seconds using setTimeout workaround
    await new Promise((resolve) => setTimeout(resolve, 5000)); // wait 5 seconds
    console.log("Step 2: Waited for 5 seconds.");

    // Try to find the specific div by id
    const specificDiv = await page.$("div[id='#/fx/nty+TdqV64UxDP81WA==']");
    if (specificDiv) {
      console.log("Step 3: Found div with id #/fx/nty+TdqV64UxDP81WA==.");
      const divContent = await page.evaluate((el) => el.innerHTML, specificDiv);
      console.log("Div content:", divContent); // Log the content of the specific div
    } else {
      console.log("Step 3: Div with id #/fx/nty+TdqV64UxDP81WA== not found.");
    }

    // You can also try to wait for the specific div with this exact id
    await page.waitForSelector("div[id='#/fx/nty+TdqV64UxDP81WA==']", {
      timeout: 15000,
    });
    console.log(
      "Step 4: Found specific div with id #/fx/nty+TdqV64UxDP81WA==."
    );

    // Proceed with scraping profiles
    const profiles = await page.evaluate(() => {
      const container = document.querySelector("div[id^='#/fx/']");
      if (!container) return [];

      const items = container.querySelectorAll("div > ul > li");

      return Array.from(items).map((li) => {
        const name =
          li.querySelector("span[aria-hidden='true']")?.innerText || "N/A";
        const title =
          li.querySelector("div.t-14.t-black.t-normal")?.innerText || "N/A";
        const location =
          li.querySelector("div.t-14.t-normal:nth-of-type(2)")?.innerText ||
          "N/A";
        const profileUrl =
          li.querySelector("a.nuXDIvMbeMYWApPugutCOKmVhZzvTYUM")?.href || "N/A";

        return {
          full_name: name.trim(),
          job_title: title.trim(),
          location: location.trim(),
          profile_url: profileUrl.trim(),
        };
      });
    });

    console.log("✅ Scraped Profiles:", profiles.length);

    // Save to MongoDB
    await client.connect();
    const db = client.db(dbName);
    const collection = db.collection("linkedin_profiles");

    for (const profile of profiles) {
      await collection.updateOne(
        { profile_url: profile.profile_url },
        { $set: profile },
        { upsert: true }
      );
    }

    console.log("📦 Profiles saved to MongoDB.");
  } catch (err) {
    console.error("❌ Error during scraping:", err);
  } finally {
    await client.close();
    await browser.close();
  }
};

scrapeLinkedIn();

File: scraper.py
import asyncio
import nodriver as uc
from pymongo import MongoClient
from datetime import datetime

# MongoDB Configuration
MONGO_URI = "mongodb+srv://mavinash422:cCRAQrT8blgY5fWf@cluster0.bic32gr.mongodb.net/Outflo?retryWrites=true&w=majority&appName=Cluster0"
DB_NAME = "Outflo"
COLLECTION_NAME = "linkedin_profiles"

async def log(message, level="INFO"):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] [{level}] {message}")

async def scrape_linkedin():
    # Initialize MongoDB client
    try:
        mongo_client = MongoClient(MONGO_URI)
        db = mongo_client[DB_NAME]
        collection = db[COLLECTION_NAME]
        await log("✅ MongoDB client initialized.")
    except Exception as e:
        await log(f"❌ Failed to connect to MongoDB: {e}", "ERROR")
        return

    # Launch browser with user persistence
    try:
        await log("🚀 Launching browser with user persistence...")
        browser = await uc.start(
            headless=False,  # Set to True in production
            user_data_dir="./linkedin_user_data",  # Persists cookies/sessions
            slow_mo=0.05,
            args=[
                "--disable-blink-features=AutomationControlled",
                "--no-sandbox",
                "--disable-setuid-sandbox",
            ],
        )
        await log("🌐 Browser launched successfully.")
    except Exception as e:
        await log(f"❌ Browser launch failed: {e}", "ERROR")
        mongo_client.close()
        return

    try:
        # Open a new tab
        page = await browser.get("about:blank")
        await log("📄 New tab opened.")

        # LinkedIn search URL
        search_url = "https://www.linkedin.com/search/results/people/?geoUrn=%5B%22103644278%22%5D&industry=%5B%221594%22%2C%221862%22%2C%2280%22%5D&keywords=%22lead%20generation%20agency%22&origin=GLOBAL_SEARCH_HEADER&sid=z%40k&titleFreeText=Founder"

        # Navigate to LinkedIn
        await log(f"🔍 Navigating to LinkedIn search URL: {search_url}")
        await page.goto(search_url, timeout=60000)
        await log("✅ Page loaded.")

        # Check if logged in (look for the LinkedIn feed icon)
        try:
            await page.wait_for_selector("nav.global-nav", timeout=10000)
            await log("🔑 User is logged in (persistence worked).")
        except:
            await log("⚠️ Not logged in. Manual login may be required.", "WARNING")

        # Wait for dynamic content
        await log("⏳ Waiting for dynamic content (5 seconds)...")
        await asyncio.sleep(5)

        # Debug: Capture screenshot (optional)
        await page.screenshot(path="./debug_screenshot.png")
        await log("📸 Screenshot saved for debugging.")

        # Scrape profiles
        await log("🕵️‍♂️ Scraping profiles...")
        profiles = await page.evaluate(
            """
            () => {
                const results = [];
                const items = document.querySelectorAll("div.entity-result__item");
                
                items.forEach((item) => {
                    const nameElem = item.querySelector("span.entity-result__title-text a span[aria-hidden]");
                    const titleElem = item.querySelector("div.entity-result__primary-subtitle");
                    const locationElem = item.querySelector("div.entity-result__secondary-subtitle");
                    const profileUrlElem = item.querySelector("a.app-aware-link");
                    
                    results.push({
                        full_name: nameElem?.innerText.trim() || "N/A",
                        job_title: titleElem?.innerText.trim() || "N/A",
                        location: locationElem?.innerText.trim() || "N/A",
                        profile_url: profileUrlElem?.href.split('?')[0] || "N/A",  // Clean URL
                    });
                });
                return results;
            }
            """
        )

        await log(f"✅ Found {len(profiles)} profiles.")

        # Save to MongoDB
        if profiles:
            await log("💾 Saving profiles to MongoDB...")
            for profile in profiles:
                try:
                    await collection.update_one(
                        {"profile_url": profile["profile_url"]},
                        {"$set": profile},
                        upsert=True,
                    )
                except Exception as e:
                    await log(f"❌ Failed to save profile {profile['full_name']}: {e}", "WARNING")
            await log("📦 All profiles saved successfully.")
        else:
            await log("⚠️ No profiles found. Check selectors or LinkedIn structure.", "WARNING")

    except Exception as e:
        await log(f"❌ Critical error during scraping: {e}", "ERROR")
    finally:
        await log("🧹 Closing resources...")
        await browser.close()
        mongo_client.close()
        await log("🎉 Scraping session completed.")

# Run the scraper
if __name__ == "__main__":
    asyncio.run(scrape_linkedin())